File: cpp/bob/bob.cpp
#include "bob.h"

string bob::hey(string s) {
    unsigned long l = s.length();
    if (isSilence(s)) {
        return "Fine. Be that way!";
    }
    else if (onlyNumbers(s)) {
        if (isQuestion(s)){
            return "Sure.";
        }
        return "Whatever.";
    }
    else if (isYelling(s)) {
        return "Whoa, chill out!";
    }
    else if (isQuestion(s)) {
        return "Sure.";
    }
    return "Whatever.";
}

bool bob::onlyNumbers(string s) {
    for(char& c : s) {
        if (isalpha(c)) {
            return false;
        }
    }
    return true;
}

bool bob::isQuestion(string s) {
    for (string::size_type i = s.length() - 1; i >= 0; i--) {
        if (!isblank(s[i])) {
            return s[i] == '?';
        }
    }
    return false;
}

bool bob::isSilence(string s) {
    for(char& c : s) {
        if (!isblank(c)) {
            return false;
        }
    }
    return true;
}

bool bob::isYelling(string s) {
    for(char& c : s) {
            if (islower(c)) {
                return false;
            }
    }
    return true;
}

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where
import Data.Char (toUpper, isDigit, isAlpha, isSpace)

responseFor :: String -> String
responseFor s = responseFor' (filter (not . isSpace) s)
    where 
    responseFor' s | isSilence s   = "Fine. Be that way!"
                   | hasNoWords s  = "Whatever."              
                   | isYelling s   = "Whoa, chill out!"
                   | isAsking s    = "Sure."              
                   | otherwise     = "Whatever."
        where
        isSilence s  = s == []
        isYelling s  = filter isAlpha s /= [] && s == map toUpper s 
        isAsking s   = last s == '?' && not (isYelling s)
        hasNoWords s = filter isAlpha s == [] && not (isAsking s)
    
